### DISCUSSION TITLE
Splitting ESTree types into separate library

### DISCUSSION THREAD

**Discussion by lahma** [adams85/acornima#5] [Ideas]
Just an idea. What if there were a (source generator) NuGet package that would hold just the EStree models so that both Esprima and Acornima could use the same models? As ESTree is a spec the models should be quite interchangeable between the parsers too. With source generator project there would be no new dependency to carry, but the sources would be part of the main library.

Would make switching between the two quite easy, even though Esprima's future doesn't look that bright with this new competitor in town 😉

**Comment by adams85** [comment]
While I see what good this would bring to the table, I doubt that it could work in practice. A common AST model would introduce an interdependency between Esprima and Acornima, which would create new problems that both parsers would need to deal with. Some gotchas off the top of my head:
* Although Acornima copied the Esprima AST, it has already made a bunch of changes/improvements to it. So, if we want a common AST, we either need to revert these changes (which, of course, wouldn't make much sense), or port them back to Esprima. I don't really have the time and willingness to do that. But even if had, it would have a ripple effect: Jint and other consumers of Esprima would need to be adjusted too...
* I'm pretty sure that the two parsers can't/won't really have the exact some models at specific points in time. For example,
  * Acorn(ima) can preserve parenthesized expressions and it needs a non-standard AST node type for representing such nodes. Esprima doesn't support this currently, so it would be weird to have a `VisitParenthesizedExpression` in its visitors.
  * Another similar issue: the two parsers will certainly be developed at different speeds, new JS language features will land at different times. Which means that they'll need different subsets of the EStree spec at given times.
  Of course, we can work around this issue by making the source generator parameterizable, however implementing and maintaining that would definitely need some extra effort too.

Anyway, I don't want to kill the idea immediately, as it's a nice one fundamentally. However, to make it a realistic one, we'd need to standardize the AST model (beyond the EStree spec, as it doesn't specify some details and C#-specific aspects) and would also need someone who takes care of the Esprima side.

**Reply by lahma** [reply]
Thank you for the feedback, probably something to consider when there's nothing left in the world to do then 😅
